cpmpy/transformations/linearize.py

```diff
@@ -5,190 +5,183 @@
 from ..expressions.globalconstraints import GlobalConstraint
-from ..expressions.utils import is_any_list, is_num
+from ..expressions.utils import is_any_list, is_num, eval_comparison, is_bool
+
 from ..expressions.variables import _BoolVarImpl, boolvar, NegBoolView, _NumVarImpl
 
-def linearize_constraint(cpm_expr):
+def linearize_constraint(cpm_expr, supported={"sum","wsum"}):
     """
     Transforms all constraints to a linear form.
     This function assumes all constraints are in 'flat normal form'.
     Only apply after 'cpmpy.transformations.flatten_model.flatten_constraint()'.
     """
 
     if is_any_list(cpm_expr):
-        lin_cons = [linearize_constraint(expr) for expr in cpm_expr]
+        lin_cons = [linearize_constraint(expr, supported=supported) for expr in cpm_expr]
         return [c for l in lin_cons for c in l]
 
+    # boolvar
     if isinstance(cpm_expr, _BoolVarImpl):
         if isinstance(cpm_expr, NegBoolView):
-            return [cpm_expr._bv <= 0]
-        return [cpm_expr >= 1]
+            return [sum([cpm_expr._bv]) <= 0]
+        return [sum([cpm_expr]) >= 1]
 
+    # conjunction
     if cpm_expr.name == "and":
-        if all(arg.is_bool() for arg in cpm_expr.args):
-            return [sum(cpm_expr.args) >= len(cpm_expr.args)]
-        raise Exception("Numeric constants or numeric variables not allowed as base constraint")
+        return [sum(cpm_expr.args) >= len(cpm_expr.args)]
 
+    # disjunction
     if cpm_expr.name == "or":
-        if all(arg.is_bool() for arg in cpm_expr.args):
-            return [sum(cpm_expr.args) >= 1]
-        raise Exception("Numeric constants or numeric variables not allowed as base constraint")
+        return [sum(cpm_expr.args) >= 1]
 
+    # reification
     if cpm_expr.name == "->":
+        # determine direction of implication
         cond, sub_expr = cpm_expr.args
-        if not cond.is_bool() or not sub_expr.is_bool():
-            raise Exception(
-                f"Numeric constants or numeric variables not allowed as base constraint, cannot linearize {cpm_expr}")
-        lin_comps = linearize_constraint(sub_expr)
-        lin_exprs = []
-        for l_expr in lin_comps:
-            lhs, rhs = l_expr.args
-            if lhs.name == "mul" and _wsum_should(lhs):
-                # multiplication with var and constant, convert to wsum
-                lhs = Operator("wsum", _wsum_make(lhs))
 
-            if isinstance(rhs, _NumVarImpl):
-                # Vars on rhs of linexpr not supported in indicator constraints
-                if isinstance(lhs, _NumVarImpl):
-                    new_lhs = lhs + (-1) * rhs
-                elif lhs.name == "sum":
-                    new_lhs = Operator("wsum", [[1]*len(lhs.args) + [-1], lhs.args + [rhs]])
-                elif lhs.name == "wsum":
-                    new_lhs = lhs + (-1 * rhs)
+        if isinstance(cond, _BoolVarImpl) and isinstance(sub_expr, _BoolVarImpl):
+            # shortcut for BV -> BV exprs
```

Comment:
this can be more uniform/explicit, that BV1 -> BV2 equals ~BV1 | BV2 equals ((1-BV1) + BV2 >= 1) where in case of BV being a negboolview an additional NBV=(1-BV) is applied.
but then I guess you purposefully already group the consts? Can be a bit more expl in docs still : )